lib/volk_prefs.c

```diff
@@ -63,6 +64,48 @@ void volk_get_config_path(char* path, bool read)
     return;
 }
 
+
+static struct volk_preferences {
+    volk_arch_pref_t* volk_arch_prefs;
+    size_t n_arch_prefs;
+    atomic_int initialized;
+
+} volk_preferences;
+
+
+void volk_initialize_preferences()
+{
+    if (!atomic_fetch_and(&volk_preferences.initialized, 1)) {
+        volk_preferences.n_arch_prefs =
+            volk_load_preferences(&volk_preferences.volk_arch_prefs);
+    }
+}
+
+
+void volk_free_preferences()
+{
+    if (volk_preferences.initialized) {
+        free(volk_preferences.volk_arch_prefs);
+        volk_preferences.n_arch_prefs = 0;
+        volk_preferences.initialized = 0;
+    }
+}
```

Comment:
This is not thread safe if two threads call the function concurrently: Thread 1 can observe `initialized` being true and consequently enter cleanup, while Thread 2 is already in the process of cleaning up (i.e. Thread 2 is somewhere between “comparison just finished” and “just before resetting `initialized`” at the time point Thread 1 performs the comparison). This issue could be “circumvented” by stating that this function is not threadsafe (i.e. its up to the used to do the right thing).

Reply:
Totally agree. If we figure out how to add a mutex to `volk_preferences` I'd just use this mutex here as well.